crypt.go

```diff
@@ -493,3 +508,136 @@ func ActivateVolumeWithRecoveryKey(volumeName, sourceDevicePath string, keyReade
 
 	return activateWithRecoveryKey(volumeName, sourceDevicePath, keyReader, options.Tries, RecoveryKeyUsageReasonRequested, activateOptions)
 }
+
+func setLUKS2KeyslotPreferred(devicePath string, slot int) error {
+	cmd := exec.Command("cryptsetup", "config", "--priority", "prefer", "--key-slot", strconv.Itoa(slot), devicePath)
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return osutil.OutputErr(output, err)
+	}
+
+	return nil
+}
+
+// InitializeLUKS2Container will initialize the partition at the specified devicePath as a new LUKS2 container. This can only
+// be called on a partition that isn't mapped. The label for the new LUKS2 container is provided via the label argument.
+//
+// The initial key used for unlocking the container is provided via the key argument, and must be a cryptographically secure
+// 64-byte random number. The key should be stored encrypted by using SealKeyToTPM.
+//
+// The container will be configured to encrypt data with AES-256 and XTS block cipher mode.
+//
+// On failure, this will return an error containing the output of the cryptsetup command.
+//
+// WARNING: This function is destructive. Calling this on an existing LUKS container will make the data contained inside of it
+// irretrievable.
+func InitializeLUKS2Container(devicePath, label string, key []byte) error {
+	if len(key) != 64 {
+		return fmt.Errorf("expected a key length of 512-bits (got %d)", len(key)*8)
+	}
+
+	cmd := exec.Command("cryptsetup",
+		// batch processing, no password verification for formatting an existing LUKS container
+		"-q",
+		// formatting a new volume
+		"luksFormat",
+		// use LUKS2
+		"--type", "luks2",
+		// read the key from stdin
+		"--key-file", "-",
+		// use AES-256 with XTS block cipher mode (XTS requires 2 keys)
+		"--cipher", "aes-xts-plain64", "--key-size", "512",
+		// use argon2i as the KDF with minimum cost (lowest possible time and memory costs). This is done
+		// because the supplied input key has the same entropy (512-bits) as the derived key and therefore
+		// increased time or memory cost don't provide a security benefit (but does slow down unlocking).
+		"--pbkdf", "argon2i", "--pbkdf-force-iterations", "4", "--pbkdf-memory", "32",
+		// set LUKS2 label
+		"--label", label,
+		// device to format
+		devicePath)
+	cmd.Stdin = bytes.NewReader(key[:])
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return osutil.OutputErr(output, err)
+	}
+
+	return setLUKS2KeyslotPreferred(devicePath, 0)
+}
+
+func addKeyToLUKS2Container(devicePath string, existingKey, key []byte, extraArgs []string) error {
```

Comment:
maybe extraArgs should be extraOptionArgs to make clear these come before the positional args